WooCommerce/Classes/POS/Controllers/PointOfSaleItemsController.swift

```diff
@@ -14,7 +14,7 @@ protocol PointOfSaleItemsControllerProtocol {
 class PointOfSaleItemsController: PointOfSaleItemsControllerProtocol {
     private(set) var itemListStatePublisher: any Publisher<ItemListState, Never>
     private var itemListStateSubject: PassthroughSubject<ItemListState, Never> = .init()
-    private var allItems: [POSItem] = []
+    private var allItems: [any POSDisplayableItem] = []
```

Comment:
Not quite. It's pretty complicated and I only partially understand the reasons, but the `any` here is required because `POSDisplayableItem` must be `Identifiable`. 

`Identifiable` has an associated type (the type of the `id`) which means we have to specify that this array (or other variable) containing these items could contain `any` type which matches the protocol. The compiler won't be able to determine what type it is, therefore won't be able to determine the associated types either.

Equatable conformance is different; because it has the `static ==` function, you can't have a protocol which conforms to Equatable. `POSItem` didn't previously conform to it, it just said that anything adopting `POSItem` had to be equatable. 

However, because you can have heterogenous arrays of `POSItem`, you couldn't have done the following:

```
let item1: POSItem = Product(name: "Beanie")
let item2: POSItem = Variation(name: "Beanie", colour: "Blue", size: "M")

assert(item1 == item2)
```

To do that, there would need to be a function `static func ==(lhs: POSItem, rhs: POSItem)` which could compare anything which conforms to `POSItem` to anything else that conforms to `POSItem`. That's not something we can express, because only knowing the properties that are in the `POSItem` protocol doesn't give us enough to do a full comparison.

If we implemented it as:

```
static func == (lhs: POSItem, rhs: POSItem) -> Bool {
    return lhs.name == rhs.name
}
```

... then the comparison in my example above would return `true`, but actually the Product and the Variation aren't the same.

Back to the `any` type – we're not actually taking advantage of it here. It allows us to keep the `id` an unknown type, conforming to `ObjectIdentifier`. Our displayable item specifies UUID, so we don't take advantage of that. 

I'd forgotten about that shortcut I'd taken, and looked just now at removing the `UUID` specification on the protocol. Removing it breaks the use at the view layer, as SwfitUI needs to know the type at compile time when it uses `id`, and the existential any doesn't allow that.

I think it's neater to remove all use of the existential here, as we're not actually getting any benefit from it so it just adds clutter. I had hoped to be able to use it to say `ForEach(items)` instead of having to pass the `id`, but it's not actually possible. Thanks for checking with it though!

Reply:
Done in f161a33